### ISSUE TITLE
Need for system-wide profile

### ISSUE THREAD

**Issue by csousougen** [paulmillr/encrypted-dns#22]
In Mac, Safari uses the encrypted dns, all other apps like terminal and App Store use the unencrypted dns server from network settings. 
Could you please create a profile that uses system-wide encrypted dns?

**Comment by paulmillr** [comment]
How?

**Comment by csousougen** [comment]
I’ve found these system-wide configurations but i don’t know how to import into a mobileconfig 

https://www.wwdcnotes.com/notes/wwdc20/10047

**Comment by BirdInFire** [comment]
@csousougen Actually an MDM profile is by default system wide...
During the developer conference of 2020 when they have announced this feature they have said : 

"The first way is to use a single [encrypted] DNS server as the default resolver for all apps on the system. If you provide a public [encrypted] DNS server, you can now write a network extension app that configures the system to use your server. Or, if you use Mobile Device Management to configure enterprise settings on devices, **_you can push down a profile to configure encrypted DNS settings for your networks_**," 

The code you have linked is for the dev who want to use encrypted dns for their app (or make an app to activate this dns (like adguard pro for iOS)).

**Comment by BirdInFire** [comment]
@paulmillr Since it's not a bug in .mobileconfig file I ask this issue be closed.
We cannot fix it only apple can so I vote to close it but it's up to you.

**Comment by paulmillr** [comment]
@BirdInFire I won't close these and other issues that describe Apple bugs. We can't do anything but we can make bugs visible.

**Comment by csousougen** [comment]
I can say for sure that on Mac, it is not system wide. The DIG command gets results from the DNS servers that are set in Network and not from the MDM profile. If you try "dig google.com" you will see ;; SERVER: IP address in results.

**Comment by BirdInFire** [comment]
> I can say for sure that on Mac, it is not system wide. The DIG command gets results from the DNS servers that are set in Network and not from the MDM profile. If you try "dig google.com" you will see ;; SERVER: IP address in results.

i have tested it like 10 minutes ago with nextdns to see the log on DNSserver side and it's totally system wide

**Comment by BirdInFire** [comment]
Will do a more profound testing to be sure

**Comment by BirdInFire** [comment]
I have tested with a new account and done a ping on terminal on google.lol (who does not exist) after having installed their profile (used on apple.nextdns.io) and some second after the ping (who have failed since google.lol does not exist, it appeared on the logs, tried again with .jp and worked also).

Need more search but I really think it's system wide, maybe when you modify your dns (after) having installed the profile it override the profile.

**Comment by csousougen** [comment]
@BirdInFire You are correct, with ping command it works. But it does not work with DIG command and nslookup command. 
Could you please try these commands too? Is it system wide if it does not work with any command?

**Comment by BirdInFire** [comment]
Sure will do a complete test when I'm at home (actually I have done it remotely), but it's possible that some system component are exempted by this profile (network app).

**Comment by BirdInFire** [comment]
@csousougen after some search I have found some profile who add : 

`<key>PayloadScope</key>`
`<string>System</string>`

Can you add it to your profile and do your test (seem I can't replic the problem as well as you).
And see if it solve the problem ?

If it's the case I will do a PR on all profile to add it.

**Comment by csousougen** [comment]
Hi @BirdInFire 
I added the 2 lines just above </dict> in nextdns profile. Still, dig and nslookup commands do not go through the encrypted dns (no logs in nextdns). 
However, I see in the logs, queries from all apps that run on my mac, this tells that the profile is system wide, except for these two commands (dig and nslookup).

**Comment by paulmillr** [comment]
Added BirdInFire's stuff to readme.

**Comment by quyleanh** [comment]
I also have the same problem. Cannot get profile work on macOS since on iOS it works well.

**Comment by ASweatyPenguin-1** [comment]
> Added BirdInFire's stuff to readme.

Hello @paulmillr, I'm not much of a tech person, I got linked here by Sun Knudsen Privacy Guides.  Sorry if you don't take questions from newbies...Anyway, my questions:

If @BirdInFire's stuff applies this to significantly more (app store, terminal, etc.) then why not put it in the published profile? 

I'm nervous to add it because You said there "**seems**" to be an additional option, have you changed your opinion at all on @BirdInFire's entry?

Thanks for your time, this stuff is cool!

**Comment by paulmillr** [comment]
>If @BirdInFire's stuff applies this to significantly more (app store, terminal, etc.) then why not put it in the published profile?

The flag needs a thorough investigation, which I don't have time to do right now. Thorough meaning: listening to the traffic, pros & cons etc.

**Comment by fyr77** [comment]
According to the [Configuration Profile Reference](https://developer.apple.com/business/documentation/Configuration-Profile-Reference.pdf), the flag determines if something is to be installed for the current user or the whole system - this does not sound like it should influence app behaviour (it might still do so).

Exact wording: 

> Determines if the profile should be installed for the system or the user. In many cases, it determines the location of the certificate items, such as keychains. Though it is not possible to declare different payload scopes, payloads, like VPN, may automatically install their items in both scopes if needed.
> Legal values are `System` and `User`, with `User` as the default value

**Comment by fyr77** [comment]
Based on the previous comment, it seems that the encrypted DNS settings have to be implemented and honoured by each individual app. So there is no local DNS server translating all DNS requests into DoH/DoT and sending it to the defined server, rather it is just saved as a value and apps may choose to use that value when operating. 

This also means there is no way to force any app to use encryption settings, I believe. Still, it seems as if some apps only use encrypted DNS if "System" scope is selected...

**Comment by Fail-Safe** [comment]
> Still, it seems as if some apps only use encrypted DNS if "System" scope is selected...

Do you find this to still be the case in current versions of MacOS, iOS, iPadOS, etc?

**Comment by fyr77** [comment]
@Fail-Safe Unfortunately, I no longer own any apple devices to test this with. Perhaps someone else can provide some insight.

**Comment by Homas** [comment]
> Do you find this to still be the case in current versions of MacOS, iOS, iPadOS, etc?
On recent MacOS, I see the same behavior [akasandra](https://github.com/akasandra) described above.
I didn't check the user/system scope because not sure which apps except Safari/FF actually use DoH.
kdig & dog ignore the settings.

**Comment by jabenninghoff** [comment]
From what I've found, using a `.mobileconfig` for DoH does in fact work system-wide, with some caveats:

1. Applications can choose to ignore the system-level resolver and use their own. By default, this is what Firefox does for some regions, as described in the [quad9 macOS setup guide](https://docs.quad9.net/Setup_Guides/MacOS/Big_Sur_and_later_%28Encrypted%29/).
1. Command line tools that interact with DNS won't use DoH - I checked `host`, `dig`, and `nslookup` - they will use the DNS severs set in Network (or picked up from DHCP).
1. Resolving names for other command line tools _should_ work. For example:

```sh
curl -s 'https://on.quad9.net' | grep '<title>'
    <title>Yes, you ARE using quad9</title>
```

`ping` also worked (resolved to the same IP).

I'm using a modified unsigned version of quad9's `.mobileconfig`, and currently have Google's DNS servers set in Network (8.8.8.8 and 8.8.4.4.).

When testing in other apps, I wasn't able to find one that didn't successfully resolve https://on.quad9.net, including Google Chrome. I also didn't thoroughly test (ie using Wireshark).

**Comment by Adifagos** [comment]
> [@Adifagos](https://github.com/Adifagos) You can use `PlistBuddy` to add the key. For instance:
> 
> /usr/libexec/PlistBuddy "${HOME}/Downloads/quad9-https.mobileconfig" -c "Add :\"PayloadScope\" string \"System\""

HAd you read the post, you'd have seen i've got it sorted already.. though i do appreciate the info :)

As already mentioned, this is by now a 'must', not an option. Seeing how he refuses to add this (for some reason) to his lists, even though he _has_ logged in to "update" the desciption since.. made me wonder.
So, so often that i really, really don't get "software" people. You'd think something residing in a 100% logic field would be a given for some people, but, hey, not for "software" people!
 
Anyway, long as he doesn't delete this conversation, folks have a way of managing. If they read all that wall of text ^^

**Comment by paulmillr** [comment]
We need to have the JSON source files updated. Some of them don't have correct addresses, which means some profiles were not automatically re-generated. For example, cloudflare is now correct, but opendns was not updated, because JSON file doesn't have addresses (they need to be added). More details in: https://github.com/paulmillr/encrypted-dns/pull/196#issuecomment-3857167924

Signed profiles are provided by the community, not by maintainer.

Contributions are welcome!

**Comment by paulmillr** [comment]
The repo now contains updated signed profiles and also a new infrastructure to sign profiles using TLS certificates. No need for Apple developer profiles now.
